File: avalanche-rs/crates/avalanche-utils/src/bag.rs
//! A multiset (bag) implementation with threshold support.

use std::collections::HashMap;
use std::fmt;
use std::hash::Hash;

use crate::set::Set;

/// A multiset (bag) that tracks the count of each element.
///
/// Bags support threshold tracking, which allows efficiently querying
/// which elements have been added at least a certain number of times.
///
/// # Examples
///
/// ```
/// use avalanche_utils::Bag;
///
/// let mut bag = Bag::new();
/// bag.add(1);
/// bag.add(1);
/// bag.add(2);
///
/// assert_eq!(bag.count(&1), 2);
/// assert_eq!(bag.count(&2), 1);
/// assert_eq!(bag.len(), 3);
///
/// // Set a threshold
/// bag.set_threshold(2);
/// let met = bag.threshold();
/// assert!(met.contains(&1));
/// assert!(!met.contains(&2));
/// ```
#[derive(Clone)]
pub struct Bag<T: Eq + Hash + Clone> {
    counts: HashMap<T, usize>,
    size: usize,
    threshold: usize,
    met_threshold: Set<T>,
}

impl<T: Eq + Hash + Clone> Default for Bag<T> {
    fn default() -> Self {
        Self::new()
    }
}

impl<T: Eq + Hash + Clone> Bag<T> {
    /// Creates a new empty bag.
    #[must_use]
    pub fn new() -> Self {
        Self {
            counts: HashMap::with_capacity(16),
            size: 0,
            threshold: 0,
            met_threshold: Set::new(),
        }
    }

    /// Creates a bag from an iterator of elements.
    pub fn of<I: IntoIterator<Item = T>>(iter: I) -> Self {
        let mut bag = Self::new();
        for item in iter {
            bag.add(item);
        }
        bag
    }

    /// Sets the threshold for the threshold set.
    ///
    /// Elements that have been added at least `threshold` times
    /// will be included in the threshold set.
    pub fn set_threshold(&mut self, threshold: usize) {
        if self.threshold == threshold {
            return;
        }

        self.threshold = threshold;
        self.met_threshold.clear();

        for (item, &count) in &self.counts {
            if count >= threshold {
                self.met_threshold.add(item.clone());
            }
        }
    }

    /// Adds a single element to the bag.
    pub fn add(&mut self, item: T) {
        self.add_count(item, 1);
    }

    /// Adds multiple elements to the bag.
    pub fn add_all<I: IntoIterator<Item = T>>(&mut self, iter: I) {
        for item in iter {
            self.add(item);
        }
    }

    /// Adds an element with a specific count.
    ///
    /// If `count` is 0, this is a no-op.
    pub fn add_count(&mut self, item: T, count: usize) {
        if count == 0 {
            return;
        }

        let total_count = self.counts.entry(item.clone()).or_insert(0);
        *total_count += count;
        self.size += count;

        if *total_count >= self.threshold && self.threshold > 0 {
            self.met_threshold.add(item);
        }
    }

    /// Returns the count of the given element.
    #[must_use]
    pub fn count(&self, item: &T) -> usize {
        self.counts.get(item).copied().unwrap_or(0)
    }

    /// Returns the total number of elements (including duplicates).
    #[must_use]
    pub fn len(&self) -> usize {
        self.size
    }

    /// Returns `true` if the bag is empty.
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.size == 0
    }

    /// Returns a list of unique elements in the bag.
    pub fn list(&self) -> Vec<T> {
        self.counts.keys().cloned().collect()
    }

    /// Returns true if this bag contains the same elements with the same counts as the other bag.
    pub fn equals(&self, other: &Self) -> bool {
        self.size == other.size && self.counts == other.counts
    }

    /// Returns the most common element and its count.
    ///
    /// If there's a tie, any of the tied elements may be returned.
    /// Returns `None` if the bag is empty.
    pub fn mode(&self) -> Option<(T, usize)> {
        self.counts
            .iter()
            .max_by_key(|(_, &count)| count)
            .map(|(item, &count)| (item.clone(), count))
    }

    /// Returns the set of elements that have met the threshold.
    #[must_use]
    pub fn threshold(&self) -> &Set<T> {
        &self.met_threshold
    }

    /// Returns a new bag containing only elements that satisfy the predicate.
    pub fn filter<F: Fn(&T) -> bool>(&self, predicate: F) -> Self {
        let mut new_bag = Self::new();
        for (item, &count) in &self.counts {
            if predicate(item) {
                new_bag.add_count(item.clone(), count);
            }
        }
        new_bag
    }

    /// Splits the bag into two bags based on a predicate.
    ///
    /// Returns a tuple where:
    /// - The first bag contains elements where the predicate returns `false`
    /// - The second bag contains elements where the predicate returns `true`
    pub fn split<F: Fn(&T) -> bool>(&self, predicate: F) -> (Self, Self) {
        let mut false_bag = Self::new();
        let mut true_bag = Self::new();

        for (item, &count) in &self.counts {
            if predicate(item) {
                true_bag.add_count(item.clone(), count);
            } else {
                false_bag.add_count(item.clone(), count);
            }
        }

        (false_bag, true_bag)
    }

    /// Removes all instances of an element from the bag.
    pub fn remove(&mut self, item: &T) {
        if let Some(count) = self.counts.remove(item) {
            self.size -= count;
            self.met_threshold.remove(item);
        }
    }

    /// Creates a clone of this bag.
    pub fn duplicate(&self) -> Self {
        self.clone()
    }
}

impl<T: Eq + Hash + Clone + fmt::Debug> fmt::Debug for Bag<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Bag(size={}): {{", self.size)?;
        let mut first = true;
        for (item, count) in &self.counts {
            if !first {
                write!(f, ", ")?;
            }
            write!(f, "{:?}: {}", item, count)?;
            first = false;
        }
        write!(f, "}}")
    }
}

impl<T: Eq + Hash + Clone + fmt::Display> fmt::Display for Bag<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "Bag (Size = {}):", self.size)?;
        for (item, count) in &self.counts {
            writeln!(f, "    {}: {}", item, count)?;
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_new() {
        let bag: Bag<i32> = Bag::new();
        assert!(bag.is_empty());
        assert_eq!(bag.len(), 0);
    }

    #[test]
    fn test_add() {
        let mut bag = Bag::new();
        bag.add(1);
        bag.add(1);
        bag.add(2);

        assert_eq!(bag.count(&1), 2);
        assert_eq!(bag.count(&2), 1);
        assert_eq!(bag.count(&3), 0);
        assert_eq!(bag.len(), 3);
    }

    #[test]
    fn test_add_count() {
        let mut bag = Bag::new();
        bag.add_count(1, 5);
        bag.add_count(2, 3);

        assert_eq!(bag.count(&1), 5);
        assert_eq!(bag.count(&2), 3);
        assert_eq!(bag.len(), 8);
    }

    #[test]
    fn test_add_count_zero() {
        let mut bag = Bag::new();
        bag.add_count(1, 0);
        assert!(bag.is_empty());
    }

    #[test]
    fn test_of() {
        let bag = Bag::of(vec![1, 2, 2, 3, 3, 3]);
        assert_eq!(bag.count(&1), 1);
        assert_eq!(bag.count(&2), 2);
        assert_eq!(bag.count(&3), 3);
        assert_eq!(bag.len(), 6);
    }

    #[test]
    fn test_threshold() {
        let mut bag = Bag::new();
        bag.add_count(1, 5);
        bag.add_count(2, 3);
        bag.add_count(3, 1);

        bag.set_threshold(3);
        let met = bag.threshold();

        assert!(met.contains(&1));
        assert!(met.contains(&2));
        assert!(!met.contains(&3));
    }

    #[test]
    fn test_threshold_update() {
        let mut bag = Bag::new();
        bag.set_threshold(2);
        bag.add(1);
        assert!(!bag.threshold().contains(&1));
        bag.add(1);
        assert!(bag.threshold().contains(&1));
    }

    #[test]
    fn test_mode() {
        let bag = Bag::of(vec![1, 2, 2, 3, 3, 3]);
        let (mode, count) = bag.mode().unwrap();
        assert_eq!(mode, 3);
        assert_eq!(count, 3);
    }

    #[test]
    fn test_mode_empty() {
        let bag: Bag<i32> = Bag::new();
        assert!(bag.mode().is_none());
    }

    #[test]
    fn test_filter() {
        let bag = Bag::of(vec![1, 2, 3, 4, 5]);
        let even = bag.filter(|&x| x % 2 == 0);
        assert_eq!(even.len(), 2);
        assert_eq!(even.count(&2), 1);
        assert_eq!(even.count(&4), 1);
    }

    #[test]
    fn test_split() {
        let bag = Bag::of(vec![1, 2, 3, 4, 5]);
        let (odd, even) = bag.split(|&x| x % 2 == 0);
        assert_eq!(odd.len(), 3);
        assert_eq!(even.len(), 2);
    }

    #[test]
    fn test_remove() {
        let mut bag = Bag::of(vec![1, 1, 2, 2, 2]);
        bag.remove(&2);
        assert_eq!(bag.len(), 2);
        assert_eq!(bag.count(&2), 0);
    }

    #[test]
    fn test_list() {
        let bag = Bag::of(vec![1, 2, 2, 3, 3, 3]);
        let list = bag.list();
        assert_eq!(list.len(), 3);
    }

    #[test]
    fn test_equals() {
        let bag1 = Bag::of(vec![1, 2, 2]);
        let bag2 = Bag::of(vec![1, 2, 2]);
        let bag3 = Bag::of(vec![1, 2, 3]);

        assert!(bag1.equals(&bag2));
        assert!(!bag1.equals(&bag3));
    }
}

File: utils/bag/bag.go
// Copyright (C) 2019, Ava Labs, Inc. All rights reserved.
// See the file LICENSE for licensing terms.

package bag

import (
	"fmt"
	"strings"

	"golang.org/x/exp/maps"

	"github.com/ava-labs/avalanchego/utils"
	"github.com/ava-labs/avalanchego/utils/set"
)

const minBagSize = 16

// Bag is a multiset.
type Bag[T comparable] struct {
	counts map[T]int
	size   int

	threshold    int
	metThreshold set.Set[T]
}

// Of returns a Bag initialized with [elts]
func Of[T comparable](elts ...T) Bag[T] {
	var b Bag[T]
	b.Add(elts...)
	return b
}

func (b *Bag[T]) init() {
	if b.counts == nil {
		b.counts = make(map[T]int, minBagSize)
	}
}

// SetThreshold sets the number of times an element must be added to be contained in
// the threshold set.
func (b *Bag[_]) SetThreshold(threshold int) {
	if b.threshold == threshold {
		return
	}

	b.threshold = threshold
	b.metThreshold.Clear()
	for vote, count := range b.counts {
		if count >= threshold {
			b.metThreshold.Add(vote)
		}
	}
}

// Add increases the number of times each element has been seen by one.
func (b *Bag[T]) Add(elts ...T) {
	for _, elt := range elts {
		b.AddCount(elt, 1)
	}
}

// AddCount increases the number of times the element has been seen by [count].
// If [count] <= 0 this is a no-op.
func (b *Bag[T]) AddCount(elt T, count int) {
	if count <= 0 {
		return
	}

	b.init()

	totalCount := b.counts[elt] + count
	b.counts[elt] = totalCount
	b.size += count

	if totalCount >= b.threshold {
		b.metThreshold.Add(elt)
	}
}

// Count returns the number of [elt] in the bag.
func (b *Bag[T]) Count(elt T) int {
	return b.counts[elt]
}

// Len returns the number of elements in the bag.
func (b *Bag[_]) Len() int {
	return b.size
}

// List returns a list of unique elements that have been added.
// The returned list doesn't have duplicates.
func (b *Bag[T]) List() []T {
	return maps.Keys(b.counts)
}

// Equals returns true if the bags contain the same elements
func (b *Bag[T]) Equals(other Bag[T]) bool {
	return b.size == other.size && maps.Equal(b.counts, other.counts)
}

// Mode returns the most common element in the bag and the count of that element.
// If there's a tie, any of the tied element may be returned.
func (b *Bag[T]) Mode() (T, int) {
	var (
		mode     T
		modeFreq int
	)
	for elt, count := range b.counts {
		if count > modeFreq {
			mode = elt
			modeFreq = count
		}
	}

	return mode, modeFreq
}

// Threshold returns the elements that have been seen at least threshold times.
func (b *Bag[T]) Threshold() set.Set[T] {
	return b.metThreshold
}

// Returns a bag with the elements of this bag that return true for [filterFunc],
// along with their counts.
// For example, if X is in this bag with count 5, and filterFunc(X) returns true,
// then the returned bag contains X with count 5.
func (b *Bag[T]) Filter(filterFunc func(T) bool) Bag[T] {
	newBag := Bag[T]{}
	for vote, count := range b.counts {
		if filterFunc(vote) {
			newBag.AddCount(vote, count)
		}
	}
	return newBag
}

// Returns:
// 1. A bag containing the elements of this bag that return false for [splitFunc].
// 2. A bag containing the elements of this bag that return true for [splitFunc].
// Counts are preserved in the returned bags.
// For example, if X is in this bag with count 5, and splitFunc(X) is false,
// then the first returned bag has X in it with count 5.
func (b *Bag[T]) Split(splitFunc func(T) bool) [2]Bag[T] {
	splitVotes := [2]Bag[T]{}
	for vote, count := range b.counts {
		if splitFunc(vote) {
			splitVotes[1].AddCount(vote, count)
		} else {
			splitVotes[0].AddCount(vote, count)
		}
	}
	return splitVotes
}

// Remove all instances of [elt] from the bag.
func (b *Bag[T]) Remove(elt T) {
	count := b.counts[elt]
	delete(b.counts, elt)
	b.size -= count
}

func (b *Bag[T]) PrefixedString(prefix string) string {
	sb := strings.Builder{}

	sb.WriteString(fmt.Sprintf("Bag[%T]: (Size = %d)", utils.Zero[T](), b.Len()))
	for elt, count := range b.counts {
		sb.WriteString(fmt.Sprintf("\n%s    %v: %d", prefix, elt, count))
	}

	return sb.String()
}

func (b *Bag[_]) String() string {
	return b.PrefixedString("")
}

func (b *Bag[T]) Clone() Bag[T] {
	var clone Bag[T]
	for id, count := range b.counts {
		clone.AddCount(id, count)
	}
	return clone
}
